Readme.md
# CubeZ

## OUTLINE

CubeZ is a platform for testing iterative solvers.


## Copyright
- Copyright (c) 2018-2020 Research Institute for Information Technology(RIIT), Kyushu University. All rights reserved.



## Prerequisite

- Cmake
- MPI library (if parallel)
- PMlib
- PAPI (Optional)
- CBrick (In case of parallel)


## How to build

### Build

~~~
$ export HOME=/hogehoge
$ mkdir build
$ cd build
$ cmake [options] ..
$ make
$ sudo make install
~~~


### Options

`-D INSTALL_DIR=` *Install_directory*
>  Specify the directory that this library will be installed. Built library is
   installed at `install_directory/lib` and the header files are placed at
   `install_directory/include`.

`-D enable_OPENMP=` {yes | no}
>  This option makes OpenMP directives effect. Default is yes.

`-D with_MPI=` {yes | no}
>  If you use an MPI library, specify `with_MPI=yes` (default).

`-D real_type=` {float | double}
>  Specify the type of floating point. If this option is omitted, the default is float.

`-D with_PM=` {*Installed_Directory* | OFF}
> Specify the directory path that PMlib is installed, or OFF.

`-D with_FAPP=` {OFF | ON}
> Specify using Fujitsu profiler.

`-D with_CBR=` *Installed_Directory*
> Specify the directory path that CBrick is installed.

`-D with_PAPI=` *Installed_Directory*
> Specify the directory path that PAPI is installed.

`-D with_SIMD=` {OFF |256|512}
> Specify SIMD length. The default is OFF. If you want to use AVX512 specify 512.

`-D with_ACC=`{off|Pascal|Volta}
> Specify using open acc directives.

`-D with_Ftrace=` (off | on)
> In the case of Aurora, if you want to use Ftrace option, specify turn on this option.

`-D enable_VectorReduction=` (on | off)
> Vectorizable reduction expression. Default is yes. This is effective for Aurora.


### Default settng
~~~
with_MPI = ON
enable_OPENMP = ON
real_type = float
with_PAPI = OFF
with_FAPP = OFF
with_SIMD = OFF
with_Ftrace = OFF
with_ACC = OFF
enable_VectorReduction = ON
~~~


## Configure Examples

`$ export CZ_HOME=hogehoge`

In following examples, assuming that TextParser, PMlib, and CBrick are installed under the CZ_HOME directory. If not, please specify applicable directory paths.


### Serial build

When even compiling without MPI, specify `-D with_CBR=OFF` to suppress linking to CBrick library.
In the case of some Intel compiler environment, please specify environment variables before compilation.
`export CC=icc CXX=icpc F90=ifort FC=ifort`


### Mac

#### gnu serial /w PMlib /wo PAPI

~~~
$ module load gcc
$ echo $CC $CXX $FC $F90
gcc-9 g++-9 gfortran-9 gfortran-9

$ cmake -DINSTALL_DIR=${HOME}/CubeZ/CZ \
-Denable_OPENMP=ON \
-Dwith_PM=${HOME}/CubeZ/PMlib \
-Dwith_SIMD=256 ..
~~~

#### PGI serial /w PMlib /wo PAPI

~~~
$ module load pgi/19.4
$ export CC=pgcc CXX=pgc++ F90=pgf90 FC=pgf90

$ cmake -DINSTALL_DIR=${HOME}/CubeZ/CZ \
-Denable_OPENMP=ON \
-Dwith_PM=${HOME}/CubeZ/PMlib_PGI \
-Dwith_SIMD=256 ..
~~~


#### INTEL serial /w PMlib /wo PAPI

~~~
$ module load intel
$ export CC=icc CXX=icpc F90=ifort FC=ifort

$ cmake -DINSTALL_DIR=${HOME}/CubeZ/CZ \
-Denable_OPENMP=ON \
-Dwith_PM=${HOME}/CubeZ/PMlib \
-Dwith_SIMD=256 ..
~~~



### ITO subsystem A/B

####  Intel serial /w PMlib /w PAPI

~~~
$ module load intel/2018
$ export CC=icc CXX=icpc F90=ifort FC=ifort

$ cmake -DINSTALL_DIR=${HOME}/CZ \
-Denable_OPENMP=ON \
-Dwith_PM=${HOME}/opt/PMlib/intel-2018_papi-gcc-4.8.5 \
-Dwith_PAPI=${HOME}/opt/PAPI/gcc-4.8.5 \
-Dwith_SIMD=256 ..
~~~

#### Intel MPI /w PMlib /w PAPI

- 2019-12-25 性能はとれるが、計算結果が不正で反復回数が増加する

~~~
$ module load intel/2018.3 openmpi/3.1.3-nocuda-intel18.3

>> PAPI-5.7.0
$ export CC=mpiicc CXX=mpiicpc F90=mpiifort FC=mpiifort
$ ./configure --prefix=${HOME}/CZ_INTL/PAPI
$ make
$ make install

>> PMlib-6.4.5
$ cmake -DINSTALL_DIR=${HOME}/CZ_INTL/PMlib \
-Denable_OPENMP=ON \
-Dwith_MPI=ON \
-Dwith_PAPI=${HOME}/CZ_INTL/PAPI \
-Denable_Fortran=no \
-Dwith_example=no \
-Dwith_OTF=no \
-Denable_PreciseTimer=yes ..

>> CBrick-1.4.3
$ cmake -DINSTALL_DIR=${HOME}/CZ_INTL/CBrick \
      -Denable_OPENMP=ON ..

>> CubeZ-1.2.9
$ cmake -DINSTALL_DIR=${HOME}/CZ_INTL/CZ \
-Denable_OPENMP=ON \
-Dwith_MPI=ON \
-Dwith_PM=${HOME}/CZ_INTL/PMlib \
-Dwith_PAPI=${HOME}/CZ_INTL/PAPI \
-Dwith_SIMD=256 \
-Dwith_CBR=${HOME}/CZ_INTL/CBrick ..
~~~

#### Fujitsu TCS serial /w PMlib /w PAPI /w FAPP

- 2020-01-07

~~~
>> PAPI-5.7.0
$ module load gcc
$ export CC=gcc F77=gfortran
$ ./configure --prefix=${HOME}/CZ_TCS/PAPI
$ make
$ make install

$ export CC=fcc CXX=FCC F90=frt FC=frt

>> PMlib-6.4.5
$ cmake -DINSTALL_DIR=${HOME}/CZ_TCS/PMlib \
-DCMAKE_TOOLCHAIN_FILE=../cmake/Toolchain_ITO_TCS.cmake \
-Denable_OPENMP=yes \
-Dwith_MPI=no \
-Denable_Fortran=no \
-Dwith_example=no \
-Dwith_PAPI=${HOME}/CZ_TCS/PAPI \
-Dwith_OTF=no \
-Denable_PreciseTimer=OFF ..

>> CubeZ-1.2.9
$ cmake -DINSTALL_DIR=${HOME}/CZ_TCS/CZ \
-DCMAKE_TOOLCHAIN_FILE=../cmake/Toolchain_ITO_TCS.cmake \
-Denable_OPENMP=ON \
-Dwith_PAPI=${HOME}/CZ_TCS/PAPI \
-Dwith_PM=${HOME}/CZ_TCS/PMlib \
-Dwith_FAPP=ON \
-Dwith_SIMD=256 ..
~~~

#### Fujitsu TCS parallel /wo PMlib /w FAPP

- 2020-01-07

~~~
>> PAPI-5.7.0
$ module load gcc
$ export CC=gcc F77=gfortran
$ ./configure --prefix=${HOME}/CZ_TCS/PAPI
$ make
$ make install

$ export CC=mpifcc CXX=mpiFCC F90=mpifrt FC=mpifrt

>> PMlib-6.4.5
$ cmake -DINSTALL_DIR=${HOME}/CZ_TCS/PMlib \
-DCMAKE_TOOLCHAIN_FILE=../cmake/Toolchain_ITO_TCS.cmake \
-Denable_OPENMP=ON \
-Dwith_MPI=ON \
-Denable_Fortran=OFF \
-Dwith_example=OFF \
-Dwith_PAPI=${HOME}/CZ_TCS/PAPI \
-Dwith_OTF=OFF \
-Denable_PreciseTimer=OFF ..

>> CBrick-1.4.3
$ cmake -DINSTALL_DIR=${HOME}/CZ_TCS/CBrick \
-DCMAKE_TOOLCHAIN_FILE=../cmake/Toolchain_ITO_TCS.cmake \
-Denable_OPENMP=ON ..

>> CubeZ-1.2.9
$ cmake -DINSTALL_DIR=${HOME}/CZ_TCS/CZ \
-DCMAKE_TOOLCHAIN_FILE=../cmake/Toolchain_ITO_TCS.cmake \
-Denable_OPENMP=ON \
-Dwith_MPI=ON \
-Dwith_PAPI=${HOME}/CZ_TCS/PAPI \
-Dwith_PM=${HOME}/CZ_TCS/PMlib \
-Dwith_FAPP=ON \
-Dwith_SIMD=256 \
-Dwith_CBR=${HOME}/CZ_TCS/CBrick ..
~~~




#### PGI OpenACC with PAPI

~~~
$ module load pgi/19.4
$ export CC=pgcc CXX=pgc++ F90=pgf90 FC=pgf90
~~~

~~~
$ cmake -DINSTALL_DIR=${HOME}/CZ \
-Dwith_MPI=no \
-Dreal_type=float \
-Denable_OPENMP=no \
-Dwith_PM=${HOME}/opt/PMlib/pgi-19.4_papi-gcc-4.8.5 \
-Dwith_SIMD=256 \
-Dwith_PAPI=${HOME}/opt/PAPI/gcc-4.8.5 \
-Dwith_ACC=Pascal \
-Dwith_CBR=OFF ..
~~~

#### PGI CPU OpenMP with PAPI

~~~
$ cmake -DINSTALL_DIR=${HOME}/CZ \
-Dwith_MPI=no \
-Dreal_type=float \
-Denable_OPENMP=yes \
-Dwith_PM=${HOME}/opt/PMlib/pgi-19.4_papi-gcc-4.8.5 \
-Dwith_SIMD=256 \
-Dwith_PAPI=${HOME}/opt/PAPI/gcc-4.8.5 \
-Dwith_ACC=OFF \
-Dwith_CBR=OFF ..
~~~




### Aurora without PAPI

~~~
$ export CC=ncc CXX=nc++ F90=nfort FC=nfort
~~~

~~~
$ cmake -DINSTALL_DIR=${HOME}/CZ \
-DCMAKE_TOOLCHAIN_FILE=../cmake/Toolchain_NEC_Aurora.cmake \
-Dwith_MPI=OFF \
-Dwith_PM=OFF \
-Dwith_CBR=OFF \
-Dwith_Ftrace=ON -Denable_VectorReduction=OFF ..
~~~



##### Note
- On Fujitsu machines(K, fx100), confirm appropriate directory path for compiler environment.
- Before building, execute following command to clean for sure. `$ make distclean`


## Building PAPI by gcc
  
  ~~~
  $ gcc --version
  gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11)
  
  $ export CC=gcc F77=gfortran
  $ cd papi-5.7.0/src
  $ ./configure --prefix=${HOME}/opt/PAPI/gcc-4.8.5
  $ make
  $ make install
  ~~~


## Building PMlib by PGI
  gccでビルドしたPAPIオブジェクトをリンクする
  
  ~~~
  $ cd PMlib-6.4.2
  $ mkdir build
  $ cd build
  $ module load pgi/19.4
  $ export CC=pgcc CXX=pgc++ F90=pgf90 FC=pgf90
  
  $ cmake -DINSTALL_DIR=${HOME}/opt/PMlib/OMP-pgi-19.4_papi-gcc-4.8.5 \
	-Denable_OPENMP=yes \
	-Dwith_MPI=no \
	-Denable_Fortran=no \
	-Dwith_example=no \
	-Dwith_PAPI=${HOME}/opt/PAPI/gcc-4.8.5 \
	-Dwith_OTF=no \
    -Denable_PreciseTimer=yes ..
  $ make
  $ make install
  ~~~


## Contributors

- Kenji Ono


### Comment of precision of Fortran
For example, the `-CcdRR8` option for fortran preprocessor convert variables, functions, and constants higher precision version in source code. Thus, functions in source code is described by floating version.

## 使い方

~~~
$ ./cz gsz_x, gsz_y, gsz_z, linear_solver, IterationMax coef [gdv_x, gdv_y, gdv_z]
$ ./cz 124 124 124 sor2sma 10000 1.5
$ ./cz 124 124 124 sor2sma_maf 10000 1.5
$ ./cz 124 124 124 pbicgstab 10000 1.5 {jacobi, psor, sor2sma, pcr}
$ ./cz 124 124 124 pbicgstab_maf 10000 1.5 {jacobi, psor, sor2sma, pcr}
$ ./cz 124 124 124 pcr 10000 1.5
$ ./cz 124 124 124 pcr_maf 10000 1.5
~~~
 - gsz_x, gsz_y, gsz_z  全計算領域の要素数
 - linear_solver        線形ソルバの指定
   - jacobi
   - sor
   - sor2sma
   - pbicgstab
   - pcr
 - IterationMax         最大反復回数
 - coef  緩和/加速係数
 - gdv_x, gdv_y, gdv_z  領域分割数の指定、指定しない場合には自動分割


ChangeLog.md
# CubeZ

## todo
BiCGSTABの境界条件は係数に織り込むFFVC方式でないと、導入できない
現状はディリクレ条件でのテストのみ

bicgの前処理性能をテスト、pcrは少ない回数でよいか？

## REVISION HISTORY

---
- 2020-01-07 Version 1.3.0
- TCS並列コンパイル
- FAPP


---
- 2020-01-04 Version 1.2.9
- プロファイリングのモード変更
  - with_PMlib={on|off}でPMlibによるプロファイリングを制御
  - with_FAPP={on|off}により、Fujitsu profilerをPMlibとは独立に指定
- ENABLE_FAPPを導入
- Toolchain_intel_F_TCS.cmake >> Toolchain_ITO_TCS.cmake
- オプションのデフォルトをOFFへ変更

---
- 2020-01-04 Version 1.2.8
- copyright 2020
- bug fix for serial >> #include <CB_SubDomain_stub.h> in cz.h
- cz_Define.hで定義していたMPIのスタブをCB_Define_stub.hへ移動


---
- 2019-12-25 Version 1.2.7
- F_TCS用のプロファイラコール, Compiler option USE_F_TCSを変更 -Nfjcex 追加 >> 未解決
- rename : CB_Define.h >> CB_Define_stub.h
- rename : CB_SubDomain.h >> CB_SubDomain_stub.h
- #ifndef DISABLE_MPIでserialのstubと切り替え
- define REAL_TYPE in cz_Define.h and remove from CB_Define_stab.h
- update cz_comm.cpp

---
- 2019-12-11 Version 1.2.6
- add -Cpp option for USE_F_TCS
- add compile script for F_TCS in Reame.md
- replace Toolchain_K.cmake to Toolchain_F_TCS.cmake
- cz_maf.f90のflop countの演算子の重複を修正

---
- 2019-8-20 Version 1.2.5
- bug_fix : pcr_rb_esa, pcr_rb_maf_esa


---
- 2019-8-14 Version 1.2.4
- bug_fix : LSOR_PCR_RB_ESA


---
- 2019-8-13 Version 1.2.3
- add priavre vars for pcr subroutines

---
- 2019-8-13 Version 1.2.2
- bug fix : pcr_rb

---
- 2019-8-13 Version 1.2.1
- Yet Another Implemantation : pcr_rb_*

---
- 2019-8-13 Version 1.2.0
- Yet Another Implemantation : SVR init

---
- 2019-8-12 Version 1.1.9
- nVidiaTools

---
- 2019-8-12 Version 1.1.8
- Aurora compiler option : static-nec

---
- 2019-8-12 Version 1.1.7
- VRtmp_Init


---
- 2019-8-11 Version 1.1.6
- modify pcr_eda / pcr_eda_maf
- modify implementation : pcr_esa/ pcr_esa_maf, pcr_rb_esa / pcr_rb_esa_maf

---
- 2019-8-11 Version 1.1.5
- vector reduction for all maf
- exclusive control between OpanMP and OpenACC

---
- 2019-8-8 Version 1.1.4
- add enable_VectorReduction option
- add printMethod(), setStrPre()

---
- 2019-8-2 Version 1.1.3
- suppress reduction(res1) at jacobi_maf_()

---
- 2019-7-19 Version 1.1.2
- pcrv_sa_maf

---
- 2019-7-19 Version 1.1.1
- pcrv_sa

---
- 2019-7-16 Version 1.1.0
- @prague
- pcrv_maf_()

---
- 2019-7-16 Version 1.0.9 
- @prague
- modify CMakeLists.txt for Aurora
- improvement for Aurora

---
- 2019-7-4 Version 1.0.8
- GPU works

---
- 2019-7-1 Version 1.0.7
- add -ta=,,,managed


---
- 2019-7-1 Version 1.0.6
- add _OPENACC macro
- add "-cpp" option for gfortran


---
- 2019-7-1 Version 1.0.5
- modify OpenACC directive and CMakefile.txt


---
- 2019-7-1 Version 1.0.4
  - OpenACC directive
  - modify AVX/AVX512 options


---
- 2019-6-27 Version 1.0.3
  - OpenACC option
  - change SIMD option
  - rm Toolchanin INTEL_SKL

---
- 2019-6-24 Version 1.0.2
  - remove ratain option


---
- 2019-6-23 Version 1.0.1
  - add  trace option
  - IVDEP

---
- 2019-6-23 Version 1.0.0
  - Aurora compiled

---
- 2019-6-23 Version 0.9.9
  - branch aurora

---
- 2019-6-20 Version 0.9.8
  - `lsor_p7` >> `pcr`
  - `lsor_pcr_kij7` >> `pcr`

---
- 2019-6-17 Version 0.9.7
  - confirm OMP behavior
  - change ITO A/B configure script in Readme.md

---
- 2019-6-16 Version 0.9.6
  - fix `lsor_p7_maf`

---
- 2019-6-10 Version 0.9.5
  - scaling for `calc_ax_maf`, `calc_rk_maf`

---
- 2019-6-9 Version 0.9.4
  - pbicgstab_maf


---
- 2019-6-9 Version 0.9.3
- add cz_maf.f90


---
- 2019-6-9 Version 0.9.2
  - add lsor_p7 to preconditioner

---
- 2019-6-9 Version 0.9.1
  - SIMD_width => SIMD_AVX512
  - Halt test6 and 7 due to SIMD intrinsic
  - lsor_pcr_kij7 : remove collapse(2) due to loop  i= , , 2
  - remove extra source
    - jacobi, psor, sor2sma, pbicgstab, lsor_p{1-7}


---
- 2019-4-8 Version 0.9.0
  - lsor_P5, P6, P7

---
- 2019-3-22 Version 0.8.9
  - lsor_pcr_kij4, 5

---
- 2019-3-21 Version 0.8.8
  - lsor_pcr_kij  omp directive modify


---
- 2019-2-18 Version 0.8.7
   - PCR
   - lsor_pcr_kij2

---
- 2018-12-31 Version 0.8.5
   - ベストケースはlsor_k2()
   - スレッドスケジュールはstatic

---
- 2018-12-31 Version 0.8.4
  - 最適化のオプション変更
  - Alignment, SIMD_width

---
- 2018-12-28 Version 0.8.3
  - テストしない関数をobsoleteへ移動

---
- 2018-12-28 Version 0.8.2
  - LSOR_J Best case

---
- 2018-12-25 Version 0.8.1
  - LSOR_J

---
- 2018-12-18 Version 0.8.0
  - tdma6_8() unroll(8)

---
- 2018-12-18 Version 0.7.9
  - tdma7()でIn-palceをやめ、d2ベクトルで出力、スワップ

---
- 2018-12-18 Version 0.7.8
  - tdma7()で受け渡し方法の別実装

---
- 2018-12-18 Version 0.7.7
  - lsor_simd4() 2-color オーダリング

---
- 2018-12-18 Version 0.7.6
  - tdma6()

---
- 2018-12-18 Version 0.7.5
  - relax_256()
  - (i,j)サンプリング > lsor_simd3()

---
- 2018-12-18 Version 0.7.4
  - relax()

---
- 2018-12-17 Version 0.7.3
  - tdma4()

---
- 2018-12-17 Version 0.7.2
  - lsor_simd3() remainder loop の削除

---
- 2018-12-17 Version 0.7.1
  - lsor_simd2() unroll(4)の実装

---
- 2018-12-16 Version 0.7.0
  - lsor_simd() peel/body/remainderのループ長調整

---
- 2018-12-16 Version 0.6.9
  - アラインしたメモリ確保

---
- 2018-12-16 Version 0.6.8
  - LSOR_SIMD()のベース実装

---
- 2018-12-16 Version 0.6.7
  - lsor_simdにk方向内側ループ実装
  - ENABLE_AVX2/AVX512
  - ケースの整理

---
- 2018-12-14 Version 0.6.6
  - MSG テスト

---
- 2018-12-14 Version 0.6.5
  - PBiCGSTAB前処理にLSOR/LJCBを実装

---
- 2018-12-13 Version 0.6.4
  - SIMD test

---
- 2018-12-10 Version 0.6.3
  - Serial 対応
  - Ljcb分割 MSF

---
- 2018-12-09 Version 0.6.2
  - SKL オプション

---
- 2018-11-23 Version 0.6.1
  - LSORのk-loopをTDMA

---
- 2018-11-23 Version 0.6.0
  - PCR

---
- 2018-11-13 Version 0.5.3
  - LSOR
  - test2 マスク付き、D, N境界条件

---
- 2018-11-13 Version 0.5.2
  - PBiCGSTAB bug fix


---
- 2018-11-13 Version 0.5.1
  - example/scripts


---
- 2018-11-13 Version 0.5.0
  - BiCGSTAB
  - 境界条件はディリクレで統一


---
- 2018-11-10 Version 0.1.0
  - Initial setup


example/tri/test4.cpp
/*
 ###################################################################################
 #
 # CubeZ
 #
 # Copyright (C) 2018-2020 Research Institute for Information Technology(RIIT), Kyushu University.
 # All rights reserved.
 #
 ###################################################################################
 */

// @file test4.cpp
// @note multi-sysytem LU

#include "cz.h"

int order_of_PM_key=0;

pm_lib::PerfMonitor PM;

/*
 
 
 D1                         D2
 |                          |
 |                          |
 |--+--+--+--......+--+--+--|
 0  1  2  3  ......  22 23 24
    <-------------------->
       Inner points
 
 BC
 <Dirichlet> ==> 係数は境界点方向はゼロで、ソース項にマイナス値
 D1 = -3.0
 D2 =  9.0
 */

//  内点の数
#define N 23

// システムの数
#define Msystem 32

/*
 N              // 解くべき方程式の次元数
 */


void set_label(const std::string label, pm_lib::PerfMonitor::Type type)
{
  bool exclusive=true;
  
  // 登録個数のチェック
  order_of_PM_key++;
  
  if ( order_of_PM_key > PM_NUM_MAX )
  {
    printf("\tThe number of labels for Performance monitor goes over limit.\n");
    exit(0);
  }
  
  // 文字数がTM_LABEL_MAX-1を超えるものはカット
  if ( strlen(label.c_str()) > TM_LABEL_MAX-1 )
  {
    printf("\tWarning: Length of timing label must be less than %d\n", TM_LABEL_MAX-1);
  }
  
  // Performance Monitorへの登録
  PM.setProperties(label, type, exclusive);
}

inline void TIMING_start(const std::string key) {
  PM.start(key);
}

inline void TIMING_stop(const std::string key, double flopPerTask=0.0, int iterationCount=1) {
  PM.stop(key, flopPerTask, (unsigned)iterationCount);
}


int main(int argc, char *argv[])
{
  //                     0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
  //REAL_TYPE a[N+2] = { 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0};
  //REAL_TYPE b[N+2] = { 0,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2,-2, 0};
  //REAL_TYPE c[N+2] = { 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0};
  //REAL_TYPE d[N+2] = {-3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-9, 9};
  // d[]はRHSと解ベクトルに利用、置換は端点以外
  // 端点には境界条件、それ以外はソース項
  
  REAL_TYPE *a=NULL, *c=NULL, *d=NULL, *w=NULL, *b=NULL;
  int n = N;
  int numThreads=0;
  
  CZ z;
  double flop = 0.0;
  
  
  PM.initialize( PM_NUM_MAX );
  PM.setRankInfo( 1 );
  std::string Parallel_str;
  if ( numThreads > 1 ) {
    Parallel_str = "OpenMP";
  }
  else {
    Parallel_str = "Serial";
  }
  PM.setParallelMode(Parallel_str, numThreads, 1);
  
  {
    using namespace pm_lib;
    
    set_label("TDMAcoupled",  PerfMonitor::CALC);
  }
  
  
  // array
  int ms = Msystem;
  
  REAL_TYPE var_type=0;
  int nx = (n+2) * ms;
  
  a  = z.czAllocR(nx, var_type);
  b  = z.czAllocR(nx, var_type);
  c  = z.czAllocR(nx, var_type);
  d  = z.czAllocR(nx, var_type);
  w  = z.czAllocR(nx, var_type);
  
  
  for (int i=2; i<=n; i++) {
    for (int m=0; m<ms; m++) {
      a[m+ms*i]=1.0;
    }
  }
  for (int i=1; i<=n; i++) {
    for (int m=0; m<ms; m++) {
      b[m+ms*i]=-2.0;
    }
  }
  for (int i=1; i<=n-1; i++) {
    for (int m=0; m<ms; m++) {
      c[m+ms*i]=1.0;
    }
  }
  
  for (int m=0; m<ms; m++) {
    d[m+ms*0] = -3.0;
    d[m+ms*1] =  3.0;
    d[m+ms*n] = -9.0;
    d[m+ms*(n+1)] = 9.0;
  }
  
  for (int i=1; i<=n; i++)
  {
    for (int m=0; m<ms; m++) {
      a[m+ms*i] /= -2.0;
      b[m+ms*i] /= -2.0;
      c[m+ms*i] /= -2.0;
      d[m+ms*i] /= -2.0;
    }
  }
  
  flop = (double)(16.0*(double)(n-1)*ms);
  
  
  // 内点の個数と、その先頭アドレス、外点の両端は境界値
  TIMING_start("TDMAcoupled");
  //z.tdma1(n, &d[1], &a[1], &c[1], &w[1]);
  //tdma_p_(&n, &d[1], &a[1], &c[1], &w[1]);
  tdma_mp_(&n, &ms, d, a, c, w);
  TIMING_stop("TDMAcoupled", flop);
  
  // ok
  //z.tdma(n, &d[1], &a[1], &b[1], &c[1], &w[1]);
  
  char str[100];
  sprintf(str, "TDMA coupled");
  PM.print(stdout, "hoge", str);
  
  
  FILE* fp;
  fp=fopen("result.txt", "w");
  
  for (int i=0; i<=n+1; i++) {
    fprintf(fp,"%2d, ", i);
    for (int m=0; m<Msystem; m++) {
      fprintf(fp,"%6.2f, ", d[m+Msystem*i]);
    }
    fprintf(fp,"\n");
  }
  
  fclose(fp);
  
  z.czDelete(w);
  z.czDelete(a);
  z.czDelete(b);
  z.czDelete(c);
  z.czDelete(d);
  
  return 0;
}
